PA193_mnemonic_Slytherin/mnemonic.py

```diff
@@ -11,159 +11,241 @@
-# - every line has exactly 1 word (no whitespaces)
-
-def __entropy2mnemonic(entropy: bytes) -> str:
-    """Convert entropy to mnemonic phrase using dictionary.
-    Currently uses 1 default dictionary with English words.
-    # TODO Should we support multiple dictionaries for various languages?
-    :rtype: str
-    :return: Mnemonic phrase
+    
+    def __init__(self, seed: bytes):
+        """Check whether provided bytes represent a valid seed.
+        :raises ValueError: on invalid parameters
+        """
+        if not isinstance(seed, bytes) or len(seed) != SEED_LEN:
+            raise ValueError('Cannot instantiate seed')
+        self.__repr__ = seed
+
+
+    def __eq__(self, other: object) -> bool:
+        """Compare seeds in constant time to prevent timing attacks.
+        :rtype: bool
+        :return: True if seeds are the same, False otherwise.
+        """
+        # > hmac.compare_digest` uses an approach designed to prevent timing
+        # > analysis by avoiding content-based short circuiting behaviour, making
+        #  > it appropriate for cryptography
+        # > https://docs.python.org/3.7/library/hmac.html#hmac.compare_digest
+        #
+        # > Note: If a and b are of different lengths, or if an error occurs,
+        # > a timing attack could theoretically reveal information about the types
+        #  > and lengths of a and b—but not their values.
+        #
+        # Type and length of seeds is known to the attacker, but not the value of expected seed.
+        if not isinstance(other, Seed):
+            return NotImplemented
+        return hmac.compare_digest(self, other)
+
+
+    def __ne__(self, other: object) -> bool:
+        """Compare seeds in constant time to prevent timing attacks.
+        :rtype: bool
+        :return: False if seeds are the same, True otherwise.
+        """
+        return not (self == other)
+
+
+class Entropy(bytes, dictionaryAccess):
+    """Class for entropy representation.
     """
-    pass
 
+    def __init__(self, entropy: bytes):
+        """Check whether provided bytes represent a valid entropy according to BIP39.
+        https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki
 
-def __mnemonic2entropy(mnemonic: str) -> bytes:
-    """Convert mnemonic phrase to entropy using dictionary.
-    Currently uses 1 default dictionary with English words.
-    # TODO Should we support multiple dictionaries for various languages?
-    :rtype: bytes
-    :return: Entropy
-    """
-    pass
+        > The mnemonic must encode entropy in a multiple of 32 bits. With more entropy security is
+        > improved but the sentence length increases. We refer to the initial entropy length as ENT.
+        > The allowed size of ENT is 128-256 bits.
+        > https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#generating-the-mnemonic
+        :raises ValueError: on invalid parameters
+        """
+        if len(entropy) not in list(range(16, 32+1, 4)):
```

Comment:
Should we check the type of entropy here?

Reply:
Well, check for `isinstance` or just pass bytes to constructor of `bytes` class which would raise exception on invalid entropy.